data/TaskDAO.js

```diff
@@ -0,0 +1,70 @@
+const Task = require('./Task');
+// In-memory implementation
+class TaskDAO {
+    constructor() {
+        this.tasks = [];
+        this.maxID = 0;
+    }
+
+    find(criteria) {
+        if(!criteria || Object.keys(criteria).length === 0) return this.tasks;
+        else {
+            return this.tasks.filter(task => this.matchesCriteria(task, criteria));
+        }
+    }
+
+    findById(id) {
+        return this.tasks.filter(task => task.id == id);
+    }
+
+    newTask(props) {
+        if(!this.validateNewTaskProps(props)) return;
+        let taskID = ++this.maxID;
+        let task = new Task(taskID, props.title, props.course, props.deadline, props.completed, props.scheduledDate);
+        this.tasks.push(task);
+        return task;
+    }
+
+    updateTask(props) {
+        if(!this.validateUpdateTaskProps(props)) return;
+        let task = this.findById(props.id)[0];
+        if(task) {
+            task.title = props.title || task.title;
+            task.course = props.course || task.course;
+            task.deadline = props.deadline || task.deadline;
+            task.completed = props.completed || task.completed;
+            task.scheduledDate = props.scheduledDate || task.scheduledDate; 
+        }
+        return task;
+    }
+
+    // might be better to just have a deleted flag?
+    deleteByID(id) {
+        let task = this.findById(id);
+        this.tasks = this.tasks.filter((task) => task.id != id);
+        return task;
+    }
+
+    //meant to be private
+    matchesCriteria(task, criteria) {
+        let matches = true;
```

Comment:
What made me look at this is that it has a variable that  gets set specifically in a loop and the comparison operator only has one equal sign. It continues through all the keys even though you only care if it gets set to false even once. You can bail out as soon as you find a property that doesn't match

```javascript
for(key in criteria) {
    if (criteria[key] != task[key]){
      return false;
    }
  }
return true;
```

if you were to process the whole set though, just for fun examples, you could use a reduce. I like reduce if you're processing a set of something and want to get one result.

```javascript
return Object.keys(criteria).reduce((matching, key)=> {
        if(matching) return criteria[key] === task[key];
      }, true);
```

or if you don't like reduce you could use the Object prototype method

```javascript
let matching = true;
      Object.entries(criteria).forEach((key, value) => {
        if(task[key] !== value) matching = false;
      });
```